plugins/kubernetes/src/kubernetes-auth-provider/AksKubernetesAuthProvider.ts

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { KubernetesAuthProvider } from './types';
+import { KubernetesRequestBody } from '@backstage/plugin-kubernetes-common';
+import { OAuthApi } from '@backstage/core-plugin-api';
+
+/**
+ *
+ * @public
+ */
+
+export class AksKubernetesAuthProvider implements KubernetesAuthProvider {
+  authProvider: OAuthApi;
+
+  constructor(authProvider: OAuthApi) {
+    this.authProvider = authProvider;
+  }
+
+  async decorateRequestBodyForAuth(
+    requestBody: KubernetesRequestBody,
+  ): Promise<KubernetesRequestBody> {
+    const aksAuthToken: string = await this.authProvider.getAccessToken();
```

Comment:
I believe you can rely on the existing microsoft auth API instead and ask for the `6dae42f8-4368-4678-94ff-3960e28e3630/user.read` scope here. That way we can avoid the addition of a new auth provider and API.

Reply:
Unfortunately the existing microsoft auth API has `User.Read` as a default scope:

https://github.com/backstage/backstage/blob/f4e4f04152818b7ce684b62cd42b07514158bcf9/packages/core-app-api/src/apis/implementations/auth/microsoft/MicrosoftAuth.ts#L44

(this scope is for the Microsoft Graph audience) If we then request a scope for the 'Azure Kubernetes Service AAD Server' (`az ad sp show --id 6dae42f8-4368-4678-94ff-3960e28e3630 | jq -r .appDisplayName`) service principal, we'll be in the same situation as #7040 -- Azure doesn't support tokens with scopes for multiple audiences and if you ask for one, you will silently be given one for whichever audience's scope is first in the list. In practice this means that the existing microsoft auth API is only capable of furnishing tokens for Microsoft Graph.

It was your comment [here](https://github.com/backstage/backstage/issues/7040#issuecomment-1253412500) that motivated us to create a new auth API for this purpose -- maybe the landscape has changed somewhat, though?

Reply:
Ah alright, had a feeling it was something like this. Are the scopes easy to distinguish? What I'm thinking might be the best solution here is to try to build this into the existing microsoft auth API instead, so that it manages two separate access tokens. So if you on the client-side request azure scopes, the graph scopes are not included, and the other way around. If you end up requesting both that'd just throw an error immediately. Bit of a prereq for that though, do you know if the refresh token supports mixed scopes, or does auth break already at that point? If not, I think a separate `azureAuthApiRef` is likely the way to go, or w/e split makes the most sense.